### ISSUE TITLE
linear search for stream costly if many streams

### ISSUE THREAD

**Issue by huitema** [private-octopus/picoquic#569]
Doh. But this is very visible when executing the "many stream" scenario. We find hot loops on the server side in:
* `picoquic_find_ready_stream`: going over the list of streams to _skip to the first non visited stream_ accounts for 15% of the total CPU
* `picoquic_prepare_required_max_stream_data_frames` accounts for 5% of the CPU
* `h3zero_find_or_create_stream`  accounts for 17% of CPU
* `picoquic_find_stream` accounts for 10% of CPU
And on the client side:
* `picoquic_find_stream` account for 6% of CPU

**Comment by bkchr** [comment]
We should use an extra list for `picoquic_find_ready_stream` to not scan all streams.

**Comment by huitema** [comment]
Yes, we could. For example, have a list of the active streams. We may also want to look at the data structures and speed up the most common accesses.

We made a decision before to visit all the streams in round robin fashion, but I think that's probably not right. It would make sense to quickly serve some streams up to their end, so as to diminish the total number more rapidly. But you has a requirement of keep a stream open for a long time e.g. for video. Maybe we should mark those "video" streams as special and give them special treatment.

**Comment by bkchr** [comment]
I still would not priorize any stream without giving the user an api to change this priorization. The specification says that no stream should be priorized.
If we just have a list of streams that want to send data and append streams to this list, if probably will not get that long and you always just need to take the first stream and don't need to search for a stream. Doing this with a linked list, should probably we no overhead at all.

**Comment by huitema** [comment]
I am concerned that the current design is falling on its knees if for some reason the traffic load has a different pattern than we expect. The "1000 streams" test that I am doing is an attempt to model that unexpected load. But you are right, it comes down to API and application expectation. I agree that things get simpler if streams are served "in order".

Is it OK to make that change? As indicated above, this would remove 15% of the current CPU load, because we would not need to skip to the first non visited stream by walking a linear list. I made that change in PR #465 to address issue #463 (entered by @bkchr  on Dec 28, 2018). I can easily roll back some of the changes in PR #465, but the issue will reappear: one big stream gets all the bandwidth, the other streams with higher stream ID just starve.

**Comment by huitema** [comment]
@bkchr I am thinking of something else -- the 17% of CPU in  "h3zero_find_or_create_stream". This static function is called when the stack asks the application for data in an active stream. The callback arguments include the ID of the stream for which data is required, but the code then has to identify the stream context, and does that with a linear search. This is demo code, and I could try to program a faster search. But this is duplicative from the work in the stack, which already identified the stream.

What if instead of just marking a stream active as in 
```
    int picoquic_mark_active_stream(picoquic_cnx_t* cnx,
        uint64_t stream_id, int is_active);
```
the application could specify some kind of stream context, as in
```
    int picoquic_mark_active_stream(picoquic_cnx_t* cnx, 
        uint64_t stream_id, int is_active, void * stream_context);
```
and the stack passed that context when asking for more data:
```
typedef int (*picoquic_stream_data_cb_fn)(picoquic_cnx_t* cnx,
    uint64_t stream_id, uint8_t* bytes, size_t length,
    picoquic_call_back_event_t fin_or_event, void* callback_ctx, void * stream_context);
```
That would allow the application to directly find the stream's context, bypass the need to search for data, and gain back the 17% CPU that I observed in the test.

Does that seem OK?

**Comment by huitema** [comment]
PR #570 implements the "stream_context" issue presented above. It does speed up the execution of the test significantly. The PR also limits the amount of printf in the demo client, which removes the corresponding overhead in measurements. We end up at this point with 1000 transactions in 150ms in debug mode on Windows, i.e. 6600 transactions per second -- before touching the stack behavior. The heavy hitting functions on the client side are:

Function Name | Total CPU [unit, %] | Self CPU [unit, %]
| ----------- | ------: | ------: |
 picoquic_find_stream | 18 (17.31%) | 18 (17.31%)
 picoquic_demo_client_find_stream | 17 (16.35%) | 17 (16.35%)
 picoquic_find_ready_stream | 12 (11.54%) | 12 (11.54%)
 picoquic_create_stream | 5 (4.81%) | 5 (4.81%)

The client is not using the active stream API, since it only writes the "get" command on each stream, when creating the stream. The logical API there would be an extension of the `picoquic_add_to_stream` API, to add a context handler that could then be used to process the response. This would remove the 16.35% overhead observed on the client side.

**Comment by huitema** [comment]
After adding the `picoquic_add_to_stream_with_ctx` API, neither the server nor the client are CPU bound in the 1000 streams test -- peak load is less than 50% of one processor on the server, negligible on the client. The processing rate is probably constrained by the congestion window. If I repeat the same experiment attempting to create 10000 streams instead of just 1000, the test last over 14 seconds, 100 times more than the previous test, thus about 66 transactions per second. Both server and client are  CPU bound, the key functions being:

Function Name | Total CPU [unit, %] | Self CPU [unit, %] | Module
*client* | |
picoquic_find_stream	3180 (37.52%)	3175 (37.46%)
picoquic_find_ready_stream | 2782 (32.83%) | 2779 (32.79%)
picoquic_prepare_required_max_stream_data_frames | 1031 (12.17%) | 1029 (12.14%)
picoquic_create_stream | 595 (7.02%) | 588 (6.94%)
picoquic_prepare_blocked_frames | 341 (4.02%) | 340 (4.01%)
*server* | |
picoquic_find_stream | 4092 (32.74%) | 4083 (32.67%)
h3zero_find_or_create_stream | 3012 (24.10%) | 2979 (23.84%)
picoquic_find_ready_stream | 2662 (21.30%) | 2647 (21.18%)
picoquic_prepare_required_max_stream_data_frames | 1523 (12.19%) | 1518 (12.15%)
picoquic_create_stream | 640 (5.12%) | 634 (5.07%)

We may debate whether "10000 streams" is a realistic scenario, or merely a denial of service attack. One plausible defense would be to limit the number of simultaneously opened streams on the server, and enforce that using the MAX_STREAMS API.

**Comment by huitema** [comment]
On the other hand, in the "DNS over QUIC" scenarios, we can easily get a long running session accumulating a large number of "old streams" -- one stream per transaction. It makes sense to keep this issue open, even after merging PR #570

**Comment by huitema** [comment]
Many of the hot spots pointed above are fixed in Pr #581, including picoquic_find_ready_stream, picoquic_prepare_required_max_stream_data_frames and h3zero_find_or_create_stream. The best approach for fixing the overhead of "picoquic_find_ready_stream" is found in issue #571. Basically, delete stream contexts from the active list when the streams are closed, so the number of active streams does not keep growing, which immediately reduces the number of streams that need to be visited. May need to couple that with server logic to only let the client open some limited number of streams, plus efficient management of "max streams" logic.

**Comment by huitema** [comment]
After the fixes in Pr #584, the server is not CPU bound. The client is, and the bulk of the client CPU time is spent in stream list management. Arguably, replacing the linked list by a splay would fix most of that. But there is also a work around: do not create so many streams at once in the first place. The tests shows 10000 transactions in 2.8 seconds. That's about 3500 transactions per second, with the server loaded at 30% or less. A better client would probably achieve 10000 tps.

On the other hand, with a better list management the client could achieve more than 10000 tps.

**Comment by huitema** [comment]
After the last set of fixes in PR #586 we achieve 32,000 transactions per second. The server does not appear to be CPU bound. It might be possible to increase the transaction rate by loosening the "max streams" flow control parameter, or maybe by implementing a self tuning mechanism, but this is a fairly low priority.
